domains/client/domain-operator/src/snap_sync.rs

```diff
@@ -0,0 +1,481 @@
+        .snap_sync_orchestrator
+        .mark_domain_snap_sync_finished();
+
+    debug!(info = ?sync_params.domain_client.info(), "Client info after successful domain snap sync.");
+
+    Ok(())
+}
+
+/// Download and return state for specified block
+async fn download_state<Block, Client, NR>(
+    header: &Block::Header,
+    client: &Arc<Client>,
+    fork_id: Option<String>,
+    network_request: &NR,
+    sync_service: &SyncingService<Block>,
+) -> Result<ImportedState<Block>, sp_blockchain::Error>
+where
+    Block: BlockT,
+    Client: HeaderBackend<Block> + ProofProvider<Block> + Send + Sync + 'static,
+    NR: NetworkRequest,
+{
+    let block_number = *header.number();
+
+    const STATE_SYNC_RETRIES: u32 = 5;
+    const LOOP_PAUSE: Duration = Duration::from_secs(20);
+    const MAX_GET_PEERS_ATTEMPT_NUMBER: usize = 30;
+
+    for attempt in 1..=STATE_SYNC_RETRIES {
+        debug!(%block_number, %attempt, "Starting state sync...");
+
+        debug!(%block_number, "Gathering peers for state sync.");
+        let mut tried_peers = HashSet::<PeerId>::new();
+
+        let mut current_get_peers_attempt = 0usize;
+        let current_peer_id = loop {
+            current_get_peers_attempt += 1;
+
+            if current_get_peers_attempt > MAX_GET_PEERS_ATTEMPT_NUMBER {
+                return Err(sp_blockchain::Error::Application(
+                    "Got empty state blocks collection for domain snap sync".into(),
+                ));
+            }
+            let all_connected_peers = sync_service
+                .peers_info()
+                .await
+                .expect("Network service must be available.");
+
+            debug!(
+                %block_number,
+                %current_get_peers_attempt,
+                ?all_connected_peers,
+                "Download state: all connected peers"
+            );
+
+            let connected_full_peers = all_connected_peers
+                .iter()
+                .filter_map(|(peer_id, info)| (info.roles.is_full()).then_some(*peer_id))
+                .collect::<Vec<_>>();
+
+            debug!(
+                current_get_peers_attempt,
+                %block_number,
+                ?tried_peers,
+                "Download state: sync peers: {:?}", connected_full_peers
+            );
+
+            let active_peers_set = HashSet::from_iter(connected_full_peers.into_iter());
+
+            if let Some(peer_id) = active_peers_set.difference(&tried_peers).next().cloned() {
+                break peer_id;
+            }
+
+            sleep(LOOP_PAUSE).await;
+        };
+
+        tried_peers.insert(current_peer_id);
+
+        let sync_engine = SnapSyncingEngine::<Block, NR>::new(
+            client.clone(),
+            fork_id.as_deref(),
+            header.clone(),
+            false,
+            (current_peer_id, block_number),
+            network_request,
+        )?;
+
+        let last_block_from_sync_result = sync_engine.download_state().await;
```

Comment:
It may be possible that the state is too large to fit into the memory and cause OOM, not an immediate issue to fix but we need to resolve it in the long term.

Reply:
There is an upstream issue for this: https://github.com/paritytech/polkadot-sdk/issues/4